internal/controller/deletion/deletion.go

```diff
@@ -0,0 +1,256 @@
+		return true, nil
+	}
+
+	objects := ctx.buildDeletionObjects(clusterInstance)
+	return ctx.DeleteObjects(clusterInstance, objects, excludeObjects)
+}
+
+// DeleteObjects deletes the ClusterInstance rendered manifests by syncWaveGroup.
+// SyncWaves are processed in descending order, that is, highest syncWaveGroup objects are first
+// deleted and the lowest syncWaveGroups are deleted last.
+// Returns: bool, error
+// bool: indicates whether the deletion of all the specified (and owned) objects are successful
+// error: error encountered during deletion attempt
+func (ctx *DeletionContext) DeleteObjects(
+	clusterInstance *v1alpha1.ClusterInstance,
+	objects, excludeObjects []ci.RenderedObject,
+) (deletionCompleted bool, err error) {
+	deletionCompleted = false
+	err = nil
+
+	patch := client.MergeFrom(clusterInstance.DeepCopy())
+	manifestsRendered := append([]v1alpha1.ManifestReference{}, clusterInstance.Status.ManifestsRendered...)
+
+	objectSyncWaveMap := ci.SyncWaveMap{}
+	if err := objectSyncWaveMap.AddObjects(objects); err != nil {
+		ctx.Logger.Warn("Incurred error(s) collecting objects for deletion", zap.Error(err))
+	}
+
+	var (
+		deleted bool
+		errs    []error
+	)
+	defer func() {
+		// Update ManifestsRendered if there are changes
+		if !equality.Semantic.DeepEqual(clusterInstance.Status.ManifestsRendered, manifestsRendered) {
+			clusterInstance.Status.ManifestsRendered = manifestsRendered
+			if updateErr := conditions.PatchCIStatus(ctx.Context, ctx.Client, clusterInstance, patch); updateErr != nil {
+				ctx.Logger.Error("Failed to update ClusterInstance.Status.ManifestsRendered", zap.Error(updateErr))
+				errs = append(errs, updateErr)
+			}
+		}
+		err = utilerrors.NewAggregate(errs)
+	}()
+
+	ownerRefLabel := ci.GenerateOwnedByLabelValue(clusterInstance.Namespace, clusterInstance.Name)
+
+	for _, syncWave := range objectSyncWaveMap.GetDescendingSyncWaves() {
+		syncWaveGroupDeleted := true
+
+		for _, object := range objectSyncWaveMap.GetSlice(syncWave) {
+			resourceId := object.GetResourceId()
+
+			if ci.IndexOfObjectByIdentity(&object, excludeObjects) >= 0 {
+				ctx.Logger.Sugar().Debugf("Excluding object (%s) from deletion", resourceId)
+				continue
+			}
+
+			manifestRefIndex := v1alpha1.IndexOfManifestByIdentity(object.ManifestReference(), manifestsRendered)
+			if manifestRefIndex < 0 {
+				ctx.Logger.Sugar().Debugf(
+					"Skipping deletion of object (%s) not found in ClusterInstance.Status.ManifestsRendered",
+					resourceId)
+				continue
+			}
+
+			manifestRef := &manifestsRendered[manifestRefIndex]
+			deleted, err = ctx.DeleteObject(object, ownerRefLabel)
+			if deleted || (err != nil && cierrors.IsNotOwnedObject(err)) {
+				ctx.Logger.Sugar().Debugf(
+					"Removing object (%s) from ClusterInstance.Status.ManifestsRendered",
+					resourceId)
+				manifestsRendered = append(manifestsRendered[:manifestRefIndex], manifestsRendered[manifestRefIndex+1:]...)
```

Comment:
It feels like it would be simpler to maintain a list of the objects we delete then loop through the manifests at the end to create a new list that includes only what we didn't delete rather than maintaining the list on each object.

What do you think?

Reply:
100% agreed!